scrapy/extensions/feedexport.py

```diff
@@ -25,7 +25,6 @@
 from scrapy.utils.misc import create_instance, load_object
 from scrapy.utils.python import without_none_values
 
-
 logger = logging.getLogger(__name__)
```

Comment:
Hi, could you bring the empty line that was above the logger back? it should be 2 empty lines between imports and the code itself, It has now only one, thanks.

Reply:
Hi. Of course. I don't know why but any time I am trying to reformat code using PyCharm it deletes that line. This time I forgot to bring it back.

Reply:
Hey, @rafaelcapucho, I've just come across this comment and figured out PyCharm is doing the same thing to my code. I've decided to check the PEP 8 and this is what it says regarding [Blank Lines](https://www.python.org/dev/peps/pep-0008/#blank-lines):

> Surround top-level function and class definitions with two blank lines.
>
> Method definitions inside a class are surrounded by a single blank line.
>
> Extra blank lines may be used (sparingly) to separate groups of related functions. Blank lines may be omitted between a bunch of related one-liners (e.g. a set of dummy implementations).
> 
> Use blank lines in functions, sparingly, to indicate logical sections.
>
> Python accepts the control-L (i.e. ^L) form feed character as whitespace; Many tools treat these characters as page separators, so you may use them to separate pages of related sections of your file. Note, some editors and web-based code viewers may not recognize control-L as a form feed and will show another glyph in its place.

Since this is not a top-level function or class definition, I don't think it's mandatory to surround it with two blank lines. 

Do you have another source or is it something related to Scrapy development guidelines itself?

Reply:
The [convention](https://stackoverflow.com/a/2953268/939364) is to keep two blank lines between imports and the rest of the code. I think it makes sense, as you already use a single blank line to separate different groups of imports.

Reply:
Thank you for your link, @Gallaecio. 

Yet, the [top voted answer](https://stackoverflow.com/a/2953268/3072406) (marked as correct by the OP) does not state any reference:

> 1. Two blank lines between the import statements and other code.
> 2. Two blank lines between each function.

In the same line as PEP 8, [Google's Python Style Guide](https://google.github.io/styleguide/pyguide.html#35-blank-lines) also only mentions top-level function and class definitions:

> Two blank lines between top-level definitions, be they function or class definitions. One blank line between method definitions and between the class line and the first method. No blank line following a def line. Use single blank lines as you judge appropriate within functions or methods.

Let's take the `logger` variable definition, something very common across Python projects, as an example.

In Scrapy, we have some differences (both valid according to PEP 8):

- it's surrounded by two blank lines in [scrapy/downloadermiddlewares/cookies.py](https://github.com/scrapy/scrapy/blob/1d2b692666f9f5265e4845f8a30f04089515f7c5/scrapy/downloadermiddlewares/cookies.py) and in [scrapy/utils/signal.py](https://github.com/scrapy/scrapy/blob/1d2b692666f9f5265e4845f8a30f04089515f7c5/scrapy/utils/signal.py);
- but it's surrounded by only one blank line in [scrapy/spidermiddlewares/urllength.py](https://github.com/scrapy/scrapy/blob/dfbe1d95071acfbba159cea051530749b1684460/scrapy/spidermiddlewares/urllength.py) and in [scrapy/core/engine.py](https://github.com/scrapy/scrapy/blob/3d58e5e3be9e266b9b0f89b094823b76511a21b1/scrapy/core/engine.py)

Apparently, [Django is using only one blank line in this case](https://github.com/django/django/search?q=logger+%3D&unscoped_q=logger+%3D).

What I'm trying to say is that there's no reference to this convention on the [Scrapy's Coding Style](https://docs.scrapy.org/en/master/contributing.html?highlight=blank%20lines#coding-style), and IMO it's quite subjective. 

If this is important for the project to the point that we need to enforce it during code reviews, I propose we implement automatic code checks with a linter or at least document it in the Scrapy's Coding Style for future reference.

What do you think? Should we create an issue?

Reply:
> it's surrounded by two blank lines in scrapy/downloadermiddlewares/cookies.py and in scrapy/utils/signal.py

I personally like two lines after imports, those two examples are on me 😄 

I believe there are a few unwritten conventions which are not taken directly from PEP 8. For instance, @Gallaecio has been promoting alphabetically sorted imports, putting `import` lines before `from ... import` ones. I'm a big fan of [parenthesized multi-line conditions with leading logical operators](https://github.com/scrapy/scrapy/pull/4525/files#diff-115565e38381736e9e87227cb69ac0f2L64-R70).

Perhaps the time has come to write all of these conventions down.

Reply:
Sure, @elacuesta.

I've created a [ticket](https://github.com/scrapy/scrapy/issues/4654) on Scrapy to continue this discussion outside this pull request.

@rafaelcapucho, I'm sorry if I sounded like I was nitpicking stuff. I'm just trying to turn this into an objective style guideline, ideally automatically enforced by a code linter or at least well documented, independently of the number of blank lines we end up deciding to use.

Cheers!